build/tasks/AnalyzeBuildGraph.cs

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.IO;
+using System.Text;
+using System.Threading;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using NuGet.Frameworks;
+using NuGet.Versioning;
+using RepoTools.BuildGraph;
+using RepoTasks.ProjectModel;
+using RepoTasks.Utilities;
+
+namespace RepoTasks
+{
+    public class AnalyzeBuildGraph : Task, ICancelableTask
+    {
+        private readonly CancellationTokenSource _cts = new CancellationTokenSource();
+
+        /// <summary>
+        /// Repositories that we are building new versions of.
+        /// </summary>
+        [Required]
+        public ITaskItem[] Solutions { get; set; }
+
+        [Required]
+        public ITaskItem[] Artifacts { get; set; }
+
+        [Required]
+        public string Properties { get; set; }
+
+        public string StartGraphAt { get; set; }
+
+        /// <summary>
+        /// The order in which to build repositories
+        /// </summary>
+        [Output]
+        public ITaskItem[] RepositoryBuildOrder { get; set; }
+
+        public void Cancel()
+        {
+            _cts.Cancel();
+        }
+
+        public override bool Execute()
+        {
+            var packageArtifacts = GetPackageInfo(Artifacts);
+
+            var factory = new SolutionInfoFactory(Log, BuildEngine5);
+            var props = MSBuildListSplitter.GetNamedProperties(Properties);
+
+            Log.LogMessage(MessageImportance.High, $"Beginning cross-repo analysis on {Solutions.Length} solutions. Hang tight...");
+
+            var solutions = factory.Create(Solutions, props, _cts.Token);
+            Log.LogMessage($"Found {solutions.Count} and {solutions.Sum(p => p.Projects.Count)} projects");
+
+            if (_cts.IsCancellationRequested)
+            {
+                return false;
+            }
+
+            EnsureConsistentGraph(packageArtifacts.Select(p => p.Item1), solutions);
+            RepositoryBuildOrder = GetRepositoryBuildOrder(packageArtifacts, solutions.Where(s => s.ShouldBuild));
+
+            return !Log.HasLoggedErrors;
+        }
+
+        private struct VersionMismatch
+        {
+            public SolutionInfo Solution;
+            public ProjectInfo Project;
+            public string PackageId;
+            public string ActualVersion;
+            public NuGetVersion ExpectedVersion;
+        }
+
+        private void EnsureConsistentGraph(IEnumerable<PackageInfo> packages, IEnumerable<SolutionInfo> solutions)
+        {
+            // ensure versions cascade
+            var lookup = packages.ToDictionary(p => p.Id, p => p, StringComparer.OrdinalIgnoreCase);
+
+            var inconsistentVersions = new List<VersionMismatch>();
+
+            // holy crap, o^4
```

Comment:
Yes, i know. I'm going to refactor this later. I'm thinking projectmodel is basically a tree. We can probably use the visitor pattern and execute multiple visitors at once so we only have to pass over the tree once.